hack/generator/pkg/astmodel/strip_unused.go

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"k8s.io/klog/v2"
+)
+
+// TypeNameSet stores type names in no particular order without
+// duplicates.
+type TypeNameSet map[TypeName]bool
+
+// NewTypeNameSet makes a TypeNameSet containing the specified
+// names. If no elements are passed it might be nil.
+func NewTypeNameSet(initial ...TypeName) TypeNameSet {
+	var result TypeNameSet
+	for _, name := range initial {
+		result = result.Add(name)
+	}
+	return result
+}
+
+// Add includes the passed name in the set and returns the updated
+// set, so that adding can work for a nil set - this makes it more
+// convenient to add to sets kept in a map (in the way you might with
+// a map of slices).
+// TODO(babbageclunk): I like the idea of making the zero element
+// useful so a map works nicely, and the analogue with appending to a
+// slice, but the mutating and returning behaviour might be too weird.
```

Comment:
I did this and tried it out, unfortunately it makes generation on my machine go from 23s (mutable sets) to 1m31s (immutable sets) - the original with no stripping was 13s. 

Also, talking with Matt a bit more I think he's right about walking from the roots down rather than from the other types back to the root - it will get rid of the need to flip references and avoid any need for memoisation. But it would essentially end up making one big set with all the connected types, and building an immutable set of 22,000 items one at a time is going to really make the quadratic behaviour more apparent. I'll keep the changes in a branch for reference though.